mitxgraders/sampling.py

```diff
@@ -331,3 +342,243 @@ class SpecificFunctions(FunctionSamplingSet):  # pylint: disable=too-few-public-
     def gen_sample(self):
         """Return a random entry from the given list"""
         return random.choice(self.config)
+
+
+class DependentSampler(VariableSamplingSet):
+    """
+    Represents a variable that depends on other variables.
+
+    You must initialize with the list of variables this depends on as well as the formula
+    for this variable. Note that only base formulas and suffixes are available for this
+    computation.
+
+    Usage
+    =====
+    Specify a single value
+    >>> ds = DependentSampler(depends=['x', 'y', 'z'], formula="sqrt(x^2+y^2+z^2)")
+    """
+
+    # Take in an individual or tuple of numbers
+    schema_config = Schema({
+        Required('depends'): [str],
+        Required('formula'): str,
+        Required('case_sensitive', default=True): bool
+    })
+
+    def gen_sample(self):
+        """Return a random entry from the given set"""
+        raise Exception("DependentSampler must be invoked with compute_sample.")
+
+    def compute_sample(self, sample_dict):
+        """Compute the value of this sample"""
+        try:
+            result, _ = evaluator(formula=self.config['formula'],
+                                  case_sensitive=self.config['case_sensitive'],
+                                  variables=sample_dict,
+                                  functions=DEFAULT_FUNCTIONS,
+                                  suffixes=DEFAULT_SUFFIXES)
+        except CalcError:
+            raise ConfigError("Formula error in dependent sampling formula: " +
+                              self.config["formula"])
+
+        return result
+
+def gen_symbols_samples(symbols, samples, sample_from):
+    """
+    Generates a list of dictionaries mapping variable names to values.
+
+    The symbols argument will usually be config['variables']
+    or config['functions'].
+
+    Usage
+    =====
+    >>> variable_samples = gen_symbols_samples(
+    ...     ['a', 'b'],
+    ...     3,
+    ...     {
+    ...         'a': RealInterval([1,3]),
+    ...         'b': RealInterval([-4,-2])
+    ...     }
+    ... )
+    >>> variable_samples # doctest: +SKIP
+    [
+        {'a': 1.4765130193614819, 'b': -2.5596368656227217},
+        {'a': 2.3141937628942406, 'b': -2.8190938526155582},
+        {'a': 2.8169225565573566, 'b': -2.6547771579673363}
+    ]
+    """
+    # Separate independent and dependent symbols
+    independent = [
+        symbol for symbol in symbols
+        if not isinstance(sample_from[symbol], DependentSampler)
+    ]
+
+    # Generate the samples
+    sample_list = []
+    for _ in range(samples):
+        # Generate independent samples
+        sample_dict = {symbol: sample_from[symbol].gen_sample() for symbol in independent}
+
+        # Generate dependent samples, following chains as necessary
+        dependent = [
```

Comment:
I initially did not understand why this was inside the for loop (it gets mutated below). So so I moved it outside the for loop. None of the tests broke, because all tests use `samples=1`. So I would suggest changing one of the tests to having `samples=2`.

Reply:
Re API, I am also happy. Re implementation, the overall approach seems fine—multiple passes through the list, evaluating symbols if possible. One could also imagine building some sort of tree structure that reflects the variable dependencies, but this works fine. 

While trying to understand the code, I did refactored a few things that resulted in a somewhat less nesting. It also stored the dependent variables in a dictionary and deleted their keys after evaluation, so there's no need to mark evaluated symbols with empty strings. Here's my version:

```python
def is_subset(iterable, iterable_superset):
    for item in iterable:
        if item not in iterable_superset:
            return False

    return True
```
Then:
```python
# Generate dependent samples, following chains as necessary
unevaluated_dependents = {
    symbol: sample_from[symbol].config['depends'] for symbol in symbols
    if isinstance(sample_from[symbol], DependentSampler)
}
while unevaluated_dependents:
    progress_made = False
    for symbol, dependencies in unevaluated_dependents.items():
        if is_subset(dependencies, sample_dict):
            sample_dict[symbol] = sample_from[symbol].compute_sample(sample_dict)
            del unevaluated_dependents[symbol]
            progress_made = True
        else:
            continue

    if not progress_made:
        bad_symbols = ", ".join(sorted(unevaluated_dependents.keys()))
        raise ConfigError("Circularly dependent DependentSamplers detected: " + bad_symbols)
```